numpy/typing/tests/data/reveal/ufuncs.pyi

```diff
@@ -76,6 +76,16 @@ assert_type(np.matmul.identity, None)
 assert_type(np.matmul(AR_f8, AR_f8), Any)
 assert_type(np.matmul(AR_f8, AR_f8, axes=[(0, 1), (0, 1), (0, 1)]), Any)
 
+assert_type(np.vecdot.__name__, Literal["vecdot"])
+assert_type(np.vecdot.ntypes, Literal[19])
+assert_type(np.vecdot.identity, None)
+assert_type(np.vecdot.nin, Literal[2])
+assert_type(np.vecdot.nout, Literal[1])
+assert_type(np.vecdot.nargs, Literal[3])
+assert_type(np.vecdot.signature, Literal["(n),(n)->()"])
+assert_type(np.vecdot.identity, None)
+assert_type(np.vecdot(AR_f8, AR_f8), Any)
```

Comment:
There are a bunch of other tests like this that got deleted. Why only this one now? Or just an oversight?

Reply:
It would improve my confidence that the typing changes are correct if all the tests that used to be in `reveal/numeric.pyi` were reproduced here.

Reply:
Tests from `reveal/numeric.pyi` were associated with the removed typing stubs (`numpy/_core/numeric.pyi`) that I added when `vecdot` was implemented in pure Python. Later, when `vecdot` was reimplemented as a gufunc the stubs should have been removed. 

As of now `vecdot` typing stub is coming from `_GUFunc_Nin2_Nout1` only, and this: `assert_type(np.vecdot(AR_f8, AR_f8), Any)` is the only test that is valid with the current typing (same for `matmul`).
I wanted to mimic `matmul` implementation, as they are both gufuncs.

If we want to keep removed tests then there's a question how we want to proceed: Remove GUFunc based typing stub and restore `numpy/_core/numeric.pyi` typing stubs (but then the same refactoring could apply to `matmul` and others) or do we want to stick to GUFunc typing stub.

Reply:
> assert_type(np.vecdot(AR_f8, AR_f8), Any) is the only test that is valid with the current typing

Can you explain a little more why e.g. `assert_type(np.vecdot(AR_i8, AR_i8), npt.NDArray[np.signedinteger[Any]])` isn't valid? Is this a limitation of the automatically generated typing for gufuncs?

If so, it sounds like the hand-written type overrides for vecdot should be used and similar overrides should be written for matmul, or the automatically generated typing should be improved to capture the semantics of the old hand-written types. Users might write types using those in numpy 2.0 (unless you're looking to backport this to 2.0, which might not be possible given timing) so we shouldn't make the types less expressive or possibly break typing tests by changing semantics.

I'll defer to @BvB93 or others who are more experienced dealing with types about all this though.

Reply:
IIUC it's a limitation of `GUFunc` typing stub. For me we can keep `vecdot` overrides as before.

Reply:
So this is an unfortunate limitation of callables that are _not_ just normal python functions: the latter are somewhat unique in that you it is very easy the customize the signature of each individual function at the instance-level; a degree of flexibility that is in pretty much all other cases only possible at the class-level.

Unfortunately (in this context) ufuncs have a bunch of extra methods in attributes, none of which can be adequately typed with normal python functions (as was the case prior to this PR) and thus forcing the use of a single ufunc class (or small set, as we do here with these typing-only nin-/nout-based ufunc classes). 

The end result is that, unless you're willing to go through the massive task of creating a single ufunc class with customized signature for every single ufunc, that you're stuck with a less precise signature, one that is unfortunately too imprecise to actually infer the output type of the likes of `assert_type(np.vecdot(AR_i8, AR_i8), npt.NDArray[np.signedinteger[Any]])` (a limitation that unfortunately applies to _all_ ufuncs).